<a href="https://colab.research.google.com/github/gabrielasenareis/python-basico/blob/main/k_means.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [59]:
from sklearn import datasets
import numpy as np
from sklearn.metrics import confusion_matrix
from sklearn.cluster import KMeans
from sklearn.metrics import adjusted_rand_score, normalized_mutual_info_score
from sklearn.preprocessing import StandardScaler
import pandas as pd

*   Vou realizar uma tarefa de agrupamento utilizando o algoritmo 'K-means' na base de dados 'Iris', essa base de dados já contém os valores separados por classe, servindo para fazer uma comparação, posteriormente, do resultado do modelo

In [77]:
# carregamento da base de dados
iris = datasets.load_iris()
#visualização de quantos registros existem por classe
unicos, quantidade = np.unique(iris.target, return_counts=True)
print(unicos)
print(quantidade)
# existem 3 registros de classe, e 50 valores para cada um.

[0 1 2]
[50 50 50]


In [78]:
iris.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [88]:
df = pd.DataFrame(iris.data)
df

0    1    2    3
0    5.1  3.5  1.4  0.2
1    4.9  3.0  1.4  0.2
2    4.7  3.2  1.3  0.2
3    4.6  3.1  1.5  0.2
4    5.0  3.6  1.4  0.2
..   ...  ...  ...  ...
145  6.7  3.0  5.2  2.3
146  6.3  2.5  5.0  1.9
147  6.5  3.0  5.2  2.0
148  6.2  3.4  5.4  2.3
149  5.9  3.0  5.1  1.8

[150 rows x 4 columns]

In [89]:
cluster = KMeans(n_clusters = 3)
cluster.fit(df)

/usr/local/lib/python3.10/dist-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


KMeans(n_clusters=3)

In [99]:
#visualização dos centróides
centroides = cluster.cluster_centers_
centroides
# os valores "[5.006     , 3.428     , 1.462     , 0.246     ]" foram os primeiros centróides,
# onde 5.006  é a média da primeira característica para todos os pontos no cluster 1,
# 3.428 é a média da segunda característica para todos os pontos no cluster 1 e assim sucessivamente

array([[5.006     , 3.428     , 1.462     , 0.246     ],
       [5.9016129 , 2.7483871 , 4.39354839, 1.43387097],
       [6.85      , 3.07368421, 5.74210526, 2.07105263]])

In [91]:
# fazendo o agrupamento do modelo
agrupamento_clusters = cluster.labels_
agrupamento_clusters

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 2, 2, 2, 2, 1, 2, 2, 2,
       2, 2, 2, 1, 1, 2, 2, 2, 2, 1, 2, 1, 2, 1, 2, 2, 1, 1, 2, 2, 2, 2,
       2, 1, 2, 2, 2, 2, 1, 2, 2, 2, 1, 2, 2, 2, 1, 2, 2, 1], dtype=int32)

In [92]:
unicos_cluster, quantidade_cluster = np.unique(agrupamento_clusters, return_counts=True)
print(unicos_cluster)
print(quantidade_cluster)

[0 1 2]
[50 62 38]


In [93]:
# vou gerar uma matrix de contingência para fazer a comparação
resultados = confusion_matrix(iris.target, agrupamento_clusters)
resultados

array([[50,  0,  0],
       [ 0, 48,  2],
       [ 0, 14, 36]])



*   Podemos observar que de 150 valores, o modelo preditivo acertou 134.
